initial commit for eip-6384 (#6384)
ethereum/EIPs: EIPS/eip-6384.md

---
eip: 6384
title: Human-readable offline signatures
description: A method for retrieving a human-readable description of EIP-712 typed and structured data.
author: Tal Be'ery (@talbeerysec), RoiV (@DeVaz1)
discussions-to: https://ethereum-magicians.org/t/eip-6384-readable-eip-712-signatures/12752
status: Draft
type: Standards Track

> category: Interface

Reviewer: ```suggestion
category: ERC
```

created: 2023-01-08
requires: 712
---

## Abstract

This EIP introduces the `evalEIP712Buffer` function, which takes an [EIP-712](./eip-712.md) buffer and returns a human-readable text description.

## Motivation

The use case of Web3 off-chain signatures intended to be used within on-chain transaction is gaining traction and being used in multiple leading protocols (e.g. OpenSea) and standards [EIP-2612](./eip-2612.md), mainly as it offers a fee-less experience.
Attackers are known to actively and successfully abuse such off-chain signatures, leveraging the fact that users are blindly signing off-chain messages, since they are not humanly readable.
While [EIP-712](./eip-712.md) originally declared in its title that being ”humanly readable” is one of its goals, it did not live up to its promise eventually and EIP-712 messages are not understandable by an average user.

> In one example, victims that browse a malicious phishing website. It requests the victim to sign a message that will put their NFT token for sale on OpenSea platform, virtually for free.

Reviewer: Whoops, I accidentally a whole word.

```suggestion
In one example, victims browse a malicious phishing website. It requests the victim to sign a message that will put their NFT token for sale on OpenSea platform, virtually for free.
```

The user interface for some popular wallet implementations is not conveying the actual meaning of signing such transactions.

In this proposal we offer a secure and scalable method to bring true human readability to EIP-712 messages by leveraging their bound smart contracts.
As a result, once implemented this EIP wallets can upgrade their user experience from current state:

![](../assets/eip-6384/media/MiceyMask-non-compliant.png)

to a much clearer user experience:

![](../assets/eip-6384/media/ZenGo-EIP-compliant.png)

> The proposed solution solves the readability issues by allowing the wallet to query the verifyingContract. The incentives for keeping the EIP-712 message descritption as accurate as possible are alligned, as the responsibility for the description is now owned by the contract, that:

Reviewer: ```suggestion
The proposed solution solves the readability issues by allowing the wallet to query the `verifyingContract`. The incentives for keeping the EIP-712 message description as accurate as possible are aligned, as the responsibility for the description is now owned by the contract, that:
```

> - knows the message meaning exactly (and probably can reuse the code that handles this message when received on chain)

Reviewer: ```suggestion
- Knows the message meaning exactly (and probably can reuse the code that handles this message when received on chain)
```

- `domainHash` is the EIP-712's domainSeparator, a hashed `eip712Domain` struct.
- `primaryType`is the EIP-712's `primaryType`.
- `typedDataBuffer` is an ABI encoded message part of the EIP-712 full message.

### Function's output(s)

The output of the the function is an array of strings. The wallet SHOULD display them to its end-users. The wallet MAY choose to augment the returned strings with additional data. (e.g. resolve contract addresses to their name)

The strings SHOULD NOT be formatted (e.g. should not contain HTML code) and wallets SHOULD treat this string as an untrusted input and handle its rendering as such.

### Support for EIP-712 messages that are not meant to be used on-chain

If `verifyingContract` is not included in the EIP-712 domain separator, wallets MUST NOT retrieve a human-readable description using this EIP. In this case, wallets SHOULD fallback to their original EIP-712 display.

## Rationale

- We chose to implement the `typeDataBuffer` parameter as abi encoded as it is a generic way to pass the data to the contract. The alternative was to pass the `typedData` struct, which is not generic as it requires the contract to specify the message data.

> - We chose to return an array of strings and not a single string as there are potential cases where the message is composed of multiple parts. for example in the case of a multiple assets transfers in the same typedDataBuffer, the contract is advised to describe each transfer in a seperate string to allow the wallet to display each transfer separately.

Reviewer: ```suggestion
- We chose to return an array of strings and not a single string as there are potential cases where the message is composed of multiple parts. For example, in the case of a multiple assets transfers in the same `typedDataBuffer`, the contract is advised to describe each transfer in a separate string to allow the wallet to display each transfer separately.
```

---
eip: 6384
title: Humanly readable offline signatures

> description: A procedure for making a proposed to be signed buffer of typed structured data humanly readable.

Reviewer: ```suggestion
description: A method for retrieving a human-readable description of EIP-712 typed and structured data.
```

author: Tal Be'ery (@talbeerysec), RoiV (@DeVaz1)
discussions-to: https://ethereum-magicians.org/t/eip-6384-readable-eip-712-signatures/12752
status: Draft
type: Standards Track
category: Interface
created: 2023-01-08
requires: 712
---

## Abstract

> This is a standard for making a proposed to be signed typed structured data humanly readable. It includes
> 
> - Specification for including a function in the verifying contract that returns a humanly readable description of the message to be signed,
> - reference implementation,
> - Security analysis of the proposed solution

Reviewer: The abstract should be a terse summary of the Specification section, perhaps with a tiny bit of background. Since every EIP follows the same format, you don't need to give an overview of the sections.

Perhaps something like:

```suggestion
This EIP introduces the `evalEIP712Buffer` function, which takes an [EIP-712](./eip-712.md) buffer and returns a human-readable text description.
```

## Motivation

The use case of Web3 off-chain signatures intended to be used within on-chain transaction is gaining traction and being used in multiple leading protocols (e.g. OpenSea) and standards [EIP-2612](./eip-2612.md), mainly as it offers a fee-less experience.
Attackers are known to actively and successfully abuse such off-chain signatures, leveraging the fact that users are blindly signing off-chain messages, since they are not humanly readable.
While [EIP-712](./eip-712.md) originally declared in its title that being ”humanly readable” is one of its goals, it did not live up to its promise eventually and EIP-712 messages are not understandable by an average user.

> The example below shows the user exprience of victims that browse an actual phishing website. The phishing website requests the victims to sign a message that will put their NFT token for sale on OpenSea platform for a virtually zero price.
> 
> ![](../assets/eip-6384/media/MiceyMask.gif)
> 
> As this example shows, the user interface for some popular wallet implementations is not coveying the actual meaning of signing such transactions.

Reviewer: I'm not sure this gif adds much to the EIP. Perhaps just:

```suggestion
In one example, victims that browse a malicious phishing website. It requests the victim to sign a message that will put their NFT token for sale on OpenSea platform, virtually for free.

The user interface for some popular wallet implementations is not conveying the actual meaning of signing such transactions.
```

> In this proposal we offer a secure and scalable method to bring true human readability to EIP-712 messages by leveraging their binded smart contracts.

Reviewer: ```suggestion
In this proposal we offer a secure and scalable method to bring true human readability to EIP-712 messages by leveraging their bound smart contracts.
```

As a result, once implemented this EIP wallets can upgrade their user experience from current state:

![](../assets/eip-6384/media/MiceyMask-non-compliant.png)

to a much clearer user experience:

![](../assets/eip-6384/media/ZenGo-EIP-compliant.png)

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> EIP-712 already formally binds an off-chain signature to a contract, with the "verifyingContract" parameter. We suggest adding a “view” function ("stateMutability":"view") to such contracts, that returns a human readable description of the meaning of this specific off-chain buffer.

Reviewer: ```suggestion
EIP-712 already formally binds an off-chain signature to a contract, with the "verifyingContract" parameter. We suggest adding a “view” function (`"stateMutability":"view"`) to such contracts, that returns a human readable description of the meaning of this specific off-chain buffer.
```

> ```

Reviewer: ````suggestion
```solidity
````

/**
 * @dev Returns the expected result of the offchain message.
*/

>      function evalEIP712Buffer(bytes32 domainHash, string memory primaryType, bytes memory signature)

Reviewer: I would normally consider something called `signature` to be the `v`, `r`, and `s` values of an ECDSA signature. In this case, is `signature` the buffer of typed data? Probably should rename this parameter if that's the case.

Author: i've renamed the parameter name from signature to typedDataBuffer

     external
     view

>      returns (string[] memory);{

Reviewer: ```suggestion
     returns (string[] memory) {
```

   ...

}
```

**Every compliant contract MUST implement this function.**

Using this function, wallets can submit the proposed off-chain signature to the contract and present the results to the user, allowing them to enjoy an “on-chain simulation equivalent” experience to their off-chain message.

This function will have a well known name and signature, such that there is no need for updates in the EIP-712 structure.

> (Some suggested alternatives for the function name, such as "explainSignedMessage". We want to get the community feedback on the proper name)

Reviewer: ```suggestion
```

### Function's inputs

The inputs of the function:

- `domainHash` is the EIP-712's domainSeparator, a hashed `eip712Domain` struct.
- `primaryType`is the EIP-712's `primaryType`.
- `encodedSignature` is an ABI encoded message part of the EIP-712 full message.

>   We want to get the feedback of the community and specifically smart contract implementors on how they would like to receive this information: e.g. as seperate parameters as we show here or the fully ABI encoded message that includes all EIP-712 elements (including `types`).

Reviewer: ```suggestion
```

### Function's output(s)

The output of the the function is an array of strings. The wallet SHOULD display them to its end-users. The wallet MAY choose to augment the returned strings with additional data. (e.g. resolve contract addresses to their name)

The strings SHOULD NOT be formatted (e.g. should not contain HTML code) and wallets SHOULD treat this string as an untrusted input and handle its rendering as such.

> As it might be useful to provide a more structured output, we want to get the feedback of the community and specifically wallets' implementors on how they would like to receive it. One suggestion is that function would return structs that represent transfers and approvals to support a simulation like experience to the users. e.g. the `evalEIP712Buffer` function would return a strcut that represnts a transfer (contract, from, to, amount) and the wallet MAY display a resolved version of it to the end user. ("transfer 2 USDC to address 0x...")

Reviewer: ```suggestion
```

### Support for EIP-712 messages that are not meant to be used on-chain

> In case the EIP-712 message is not bounded to a contract, the wallet cannot call this function as the contract address is unkown and SHOULD default to its current display.

Reviewer: You should define what "not bound to a contract" means here. I assume you mean `verifyingContract` in the domain separator?

```suggestion
If `verifyingContract` is not included in the EIP-712 domain separator, wallets MUST NOT retrieve a human-readable description using this EIP. In this case, wallets SHOULD fallback to their original EIP-712 display.
```

> Some web2 implementations are using a signed EIP-712 message as a method to identify the user. It would be helpful if the wallet can display a reassuring message that this message cannot be used on-chain.
> We would like to get feedback from the community if this can be done safely, e.g. by omitting the `chainId` and/or `verifyingContract` from EIP-712 message, or putting non-exisiting values in them (e.g. `verifyingContract` = 0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC)

Reviewer: EIPs aren't exactly the best place for wallet feature requests ;)

```suggestion
```

## Rationale

The proposed solution solves the readability issues. The incentives for keeping the EIP-712 message descritption as accurate as possible are alligned, as the responsibility for the description is now owned by the contract, that:

- knows the message meaning exactly (and probably can reuse the code that handles this message when received on chain)
- Natively incentivized to provide the best explanation to prevent a possible fraud
- Not involving a third party that needs to be trusted
- Maintains the fee-less customer experience as the added function is in “view” mode and does not require an on-chain execution and fees.

> - Maintains Web3’s composability property

Reviewer: This is more like Motivation than Rationale. The Rationale section should describe individual technical choices made in your EIP, while the Motivation should justify the existence of the EIP as a whole.

A great example of something that should belong in your Rationale section is why your function returns an array of strings.

Author: updated accordingly :)

### Alternative solutions

#### Third party services:

Currently, the best choice for users is to rely on some 3rd party solutions that get the proposed message as input and explain its intended meaning to the user. This approach is:

- Not scalable: 3rd party provider needs to learn all such proprietary messages
- Not necessarily correct: the explanation is based on 3rd party interpretation of the original message author
- Introduces an unnecessary dependency of a third party which may have some operational, security, and privacy implications.

#### Domain name binding

> Alternatively, wallets can bind domain name to a signature. i.e. only accept EIP-712 message if it comes from a web2 domain that its `name` as defined by EIP-712 is included in `eip712Domain`. However this approach has the following disadvantages:

Reviewer: I think there's a typo here. Should `that` be `than`?

Author: should be "that", adding parentheses to make it more clear

- It breaks Web3’s composability, as now other dapps cannot interact with such messages
- Does not protect against bad messages coming from the specified web2 domain, e.g. when web2 domain is hacked

> - Some current connector, such as walletConnect do not allow wallets to verify the web2 domain authenticity

Reviewer: ```suggestion
- Some current connector, such as WalletConnect do not allow wallets to verify the web2 domain authenticity
```

## Backwards Compatibility

For non-supporting contracts the wallets will default to showing whatever they are showing today.

> Non-supporting wallets will not call this function and will default to showing whatever they are showing today.

Reviewer: Duplicate text?

Author: 1 refers to wallets, 1 refers to contracts

## Reference Implementation

A reference implementation can be found [here](../assets/eip-6384/implementation/src/MyToken/MyToken.sol).
This toy example shows how an [EIP-20](./eip-20.md) contract supporting this EIP implements an EIP-712 support for "transferWithSig" functionality (a non-standard variation on Permit, as the point of this EIP is to allow readability to non-standard EIP-712 buffers).
To illustrate the usability of this EIP to some real world use case, a helper function for the actual OpenSea's SeaPort EIP-712 is implemented too in [here](../assets/eip-6384/implementation/src/SeaPort/SeaPort712ParserHelper.sol).

## Security Considerations

### The threat model:

The attack is facilitated by a rogue web2 interface (“dapp”) that provides bad parameters for an EIP-712 formatted message that is intended to be consumed by a legitimate contract. Therefore, the message is controlled by attackers and cannot be trusted, however the contract is controlled by a legitimate party and can be trusted.

> The attacker intends to use that signed EIP-712 message on-chain later on, with a transaction crafted by the attackers. (if the subsequent on-chain transaction was to be sent by the victim, then a regular transaction simulation would suffice)

Reviewer: ```suggestion
The attacker intends to use that signed EIP-712 message on-chain later on, with a transaction crafted by the attackers. If the subsequent on-chain transaction was to be sent by the victim, then a regular transaction simulation would have sufficed.
```

The case of a rogue contract is irrelevant, as such a rogue contract can already facilitate the attack regardless of the existence of the EIP-712 formatted message.

Having said that, a rogue contract may try to abuse this functionality in order to send some maliciously crafted string in order to exploit vulnerabilities in wallet rendering of the string. Therefore wallets should treat this string as an untrusted input and handle its renderring it as such.

### Analysis of the proposed solution

The explanation is controlled by the relevant contract which is controlled by a legitimate party. The attacker must specify the relevant contract address, as otherwise it will not be accepted by it. Therefore, the attacker cannot create false explanations using this method.
Please note that if the explanation was part of the message to sign it would have been under the control of the attacker and hence irrelevant for security purposes.

Since the added functionality to the contract has the “view” modifier, it cannot change the on-chain state and harm the existing functionalities of the contract.

> 

Reviewer: I think you should still mention the attack where an attacker creates both a fake website and a fake contract, and the fake contract gives a misleading description. If the real contract's domain separator doesn't contain `verifyingContract` than the fake contract can give whatever fraudulent description it wants.

Then you can say the attack is prevented by requiring that wallets only attempt to get descriptions when the domain separator contains a verifying contract.

Author: if it's a fake contract, then it has no access to the users assets, therefore not a relevant attack scenario, hence not in scope

Reviewer: A signed message without `verifyingContract` in the domain separator would be valid for both the fake contract, and more importantly, for the real contract, right?

Reviewer 2: if there is no verifyingContract we already mentioned explicitly "In case the EIP-712 message is not bounded to a contract, the wallet cannot call this function as the contract address is unknown and SHOULD default to its current display."

Reviewer 2: Sorry the updated wording is "If verifyingContract is not included in the EIP-712 domain separator, wallets MUST NOT retrieve a human-readable description using this EIP. In this case, wallets SHOULD fallback to their original EIP-712 display."

Reviewer 2: Additionally, the signature of the rogue contract will not be valid for. the good contract, as the malicious contract buffer contains a verifyingContract in as the domain separator that would fail the signature validation for the good contract as it does not take this parameter into consideration when doing ecrecover.
